File: frogment/src/main/java/ch/grze/frogment/activity/StateAwareFrogmentActivityComponent.java
package ch.grze.frogment.activity;

import android.content.Intent;
import android.os.Bundle;

import ch.grze.frogment.ActivityStateProvider;
import ch.grze.frogment.State;
import ch.grze.frogment.StateAware;
import ch.grze.frogment.ViewStateAware;
import ch.grze.frogment.core.Core;
import ch.grze.frogment.frogment.FrogmentInterface;

final public class StateAwareFrogmentActivityComponent<S extends State> implements StateAware<S>, ViewStateAware {
    private final StateAwareFrogmentActivityInterface<S> stateAwareFrogmentActivity;
    private final Core core;

    private boolean isViewReady = false;
    private S state;

    public StateAwareFrogmentActivityComponent(Core core, StateAwareFrogmentActivityInterface<S> stateAwareFrogmentActivity) {
        this.core = core;
        this.stateAwareFrogmentActivity = stateAwareFrogmentActivity;
    }

    @Override
    public void onViewReady() {
        isViewReady = true;
        stateAwareFrogmentActivity.onViewStateChange(state);
    }

    @Override
    public boolean isViewReady() {
        return isViewReady;
    }

    @Override
    public S getState() {
        return state;
    }

    @Override
    public void setState(S state) {
        this.state = state;

        stateAwareFrogmentActivity.onBeforeStateChange(state);
        stateAwareFrogmentActivity.onStateChange(state);

        if (isViewReady) {
            stateAwareFrogmentActivity.onViewStateChange(state);
        }
    }

    @Override
    public S getDefaultState() {
        return stateAwareFrogmentActivity.getDefaultState();
    }

    public void onActivityCreated(Bundle savedInstanceState) {
        reloadState(stateAwareFrogmentActivity.getIntent(), savedInstanceState);
    }

    public void onActivityStarted() {
        onViewReady();
    }

    public void onActivitySaveInstanceState(Bundle outState) {
        stateAwareFrogmentActivity.onBeforeStateChange(state);
        outState.putParcelable(StateAwareFrogmentActivityInterface.STATE, state);
    }

    public void onFrogmentConfigure(FrogmentInterface frogment) {
        if (frogment instanceof ActivityStateProvider) {
            final ActivityStateProvider<S> provider = (ActivityStateProvider<S>) frogment;

            setState(provider.getFrogmentActivityState());
        }
    }

    private void reloadState(Intent intent, Bundle savedInstanceState) {
        final S state = core.getParser().getData(StateAwareFrogmentActivityInterface.STATE, getDefaultState(), savedInstanceState, intent);

        setState(state);
    }
}

File: frogment/src/main/java/ch/grze/frogment/activity/StateAwareFrogmentActivityComponent.kt
package ch.grze.frogment.activity

import android.content.Intent
import android.os.Bundle
import ch.grze.frogment.State
import ch.grze.frogment.StateAware
import ch.grze.frogment.StateCallbacks.*
import ch.grze.frogment.ViewStateAware
import ch.grze.frogment.core.Core
import ch.grze.frogment.core.backstack.BackStackChangeListener
import ch.grze.frogment.frogment.FrogmentInterface

class StateAwareFrogmentActivityComponent<S : State> : StateAware<S>, ViewStateAware, BackStackChangeListener {
    lateinit var core: Core
    lateinit var stateAwareFrogmentActivity: StateAwareFrogmentActivityInterface<S>

    override var isViewReady: Boolean = false

    override var state: S? = null
        set(value) {
            field = value

            state?.let {
                (stateAwareFrogmentActivity as? OnBeforeStateChange<S>)?.onBeforeStateChange(it)
                (stateAwareFrogmentActivity as? OnStateChange<S>)?.onStateChange(it)

                if (isViewReady) {
                    (stateAwareFrogmentActivity as? OnViewStateChange<S>)?.onViewStateChange(it)
                }
            }
        }

    override fun onViewReady() {
        isViewReady = true

        state?.let {
            (stateAwareFrogmentActivity as? OnViewStateChange<S>)?.onViewStateChange(it)
        }
    }

    override fun onFrogmentPushed(frogment: FrogmentInterface) {
        onFrogmentConfigure(frogment)
    }

    override fun onFrogmentPopped(frogment: FrogmentInterface) {
        onFrogmentConfigure(frogment)
    }

    override fun onBackStackEmpty() {}

    fun onActivityCreated(savedInstanceState: Bundle?) {
        reloadState(stateAwareFrogmentActivity.getIntent(), savedInstanceState)
    }

    fun onActivityStarted() {
        onViewReady()
    }

    fun onActivitySaveInstanceState(outState: Bundle) {
        state?.let {
            (stateAwareFrogmentActivity as? OnBeforeStateChange<S>)?.onBeforeStateChange(it)

            outState.putParcelable(StateAwareFrogmentActivityInterface.STATE, it)
        }
    }

    private fun onFrogmentConfigure(frogment: FrogmentInterface) {
        (frogment as? ActivityStateProvider<*>)?.let {
            state = it.frogmentActivityState as S
        }
    }

    private fun reloadState(intent: Intent, savedInstanceState: Bundle?) {
        this.state = core.parser.getData(StateAwareFrogmentActivityInterface.STATE, stateAwareFrogmentActivity.defaultState, savedInstanceState, intent)
    }
}